source/common/quic/envoy_quic_server_stream.h

```diff
@@ -32,6 +38,20 @@ class EnvoyQuicServerStream : public quic::QuicSpdyServerStreamBase,
   bool streamErrorOnInvalidHttpMessage() const override {
     return http3_options_.override_stream_error_on_invalid_http_message().value();
   }
+  // Copy this stream's StreamInfo, ensuring that the request headers pointed to by the StreamInfo
+  // lives beyond stream destruction.
+  void setDeferredLoggingHeadersAndTrailers(Http::RequestHeaderMapSharedPtr request_header_map,
+                                            Http::ResponseHeaderMapSharedPtr response_header_map,
+                                            Http::ResponseTrailerMapSharedPtr response_trailer_map,
+                                            StreamInfo::StreamInfo& stream_info) override {
+    std::unique_ptr<StreamInfo::StreamInfoImpl> new_stream_info =
+        std::make_unique<StreamInfo::StreamInfoImpl>(
+            filterManagerConnection()->dispatcher().timeSource(),
+            filterManagerConnection()->connectionInfoProviderSharedPtr());
+    new_stream_info->setFrom(stream_info, request_header_map.get());
+    stats_gatherer_->setDeferredLoggingHeadersAndTrailers(
+        request_header_map, response_header_map, response_trailer_map, std::move(new_stream_info));
```

Comment:
> If we need stream info to live longer, why don't we just make that a shared_ptr and allow it to live longer?

That was indeed the original plan, and this PR changed course in https://github.com/envoyproxy/envoy/commit/613f1c38e3fea659da394225d29335d4cef8b455. We ran into a dead end when ensuring that all of the members of the StreamInfo implementations were also shared, specifically, get/set requestHeaders in StreamInfoImpl. I believe it would have required updating a very widely used Filter interface with hundreds of call sites.

Reply:
OK. I wonder if my alternate implementation idea would somehow make this cleaner? (Have deferred logging built into the HCM so that relevant things could be moved instead of copied?)

Reply:
The reason for this codec-centric implementation is to take advantage of the QUIC AckListener interface so we don't have to deal with any of the transport layer details.

We'd like to add deferred logging for H/1 and H/2 as well, and it seems like then it might make sense to take a look at building deferred logging into HCM. In any case I'd love to hear if you have any thoughts/ideas/sketches about how it would work in HCM.

Reply:
> We'd like to add deferred logging for H/1 and H/2 as well, and it seems like then it might make sense to take a look at building deferred logging into HCM. In any case I'd love to hear if you have any thoughts/ideas/sketches about how it would work in HCM.

I would probably just have a "deferred logging manager" that is a Singleton created as part of the HCM config. For deferred logging, the relevant structures could be moved in to live longer than the HCM level connection at teardown. If we do this (move) you might actually not need most/all of these changes. Then, you could have the codec potentially register to provide callbacks to the deferred logging manager for a particular stream (and build time outs, etc. in to the  manager). In this way I think you could still mostly rely on the codec level ACK manager and to provide the callbacks.

Reply:
We are not sure if **moving** deferred logging relevant objects when HCM finishes sending response is a safe approach, in case where the moved objects gets accessed by HCM ActiveStream afterwards. So we decided to shared the ownership. I would think even if we do all the deferred logging stuff in HCM, the relevant objects currently living ActiveStream still need to be extended by shared ownership instead of moving because of the same issue. Or did I miss understand your suggestion?

Another approach is to extend the ActiveStream lifetime, but it seems non-ideal to keep all the filters alive while waiting for ACKs.

Reply:
TBH I haven't though about it / analyzed it in detail so you may be correct. This was just my gut instinct. Naively, I would think that if we defer logging until after ActiveStream goes away, we can just move things out of it? Why does it need to be shared? Basically we can move at the point we destroy ActiveStream.

Reply:
Right but it could be moved even at the point of deferred destruction. IMO moving would be simpler to reason about in this case.

Reply:
> Right but it could be moved even at the point of deferred destruction.

Do you mean moving the objects when the ActiveStream object is actually destroyed, i.e. in the destructor? That's an interesting idea! But it would need non-trival refactoring I guess. And given that ACK callbacks may be triggered before the actual destruction of ActiveStream, my gut feeling is that something still need to be shared with the codec.

Reply:
> And given that ACK callbacks may be triggered before the actual destruction of ActiveStream, my gut feeling is that something still need to be shared with the codec.

I don't think anything would need to be shared if the codec triggers the cleanup callback. If it happens before destruction, you log and then destroy. If it happens after destruction it would be handled by the "cleanup manager."

Reply:
@mattklein123 Thanks for the discussion. Are you ok with moving forward with the current implementation or do you think a change/refactor is required?

Reply:
Hmm, I would think unique ownership requires more sophisticated lifetime handling than shared ownership. Do you think we can go with this implementation for now and figure out how to shared code with other codecs and how to manage the object lifetime better when we add implementations for HTTP/2 and HTTP/1?

Reply:
I'm not thrilled about the current implementation as I think it's somewhat confusing and will have to be redone anyway once we do HTTP/1 and HTTP/2, but I'm not going to block it if you want to redo it later. I would at least add a TODO somewhere about where we want to go with this?

/wait

Reply:
Thanks Matt for the discussion here. I've added a TODO and we'll revisit this for later implementations as you suggest.